base/shell/explorer/notifyiconscust.cpp

```diff
@@ -20,26 +20,188 @@
+        SendMessageW(hCombo, CB_ADDSTRING, i, (LPARAM)szNotify.GetString());
+    }
+    return  m_hWnd;
+}
+
+BOOL TrayIconView::AddColumn(INT Index, INT Width, DWORD strId, INT fmt)
+{
+    CStringW strLabel(MAKEINTRESOURCEW(strId));
+    LVCOLUMNW Column;
+    ZeroMemory(&Column, sizeof(Column));
+
+    Column.mask =  LVCF_WIDTH | LVCF_SUBITEM | LVCF_TEXT | LVCF_FMT;
+    Column.iSubItem = Index;
+    Column.cx = Width;
+    Column.pszText = (LPWSTR)strLabel.GetString();
+    Column.cchTextMax = strLabel.GetLength();
+    Column.fmt = fmt;
+
+    return (InsertColumn(Index, &Column) == -1) ? FALSE : TRUE;
+}
+
+BOOL TrayIconView::InsertItem(LPWSTR data, INT iSubItem, INT id)
+{
+    LVITEMW iItem;
+
+    iItem.mask = LVIF_TEXT | LVIF_IMAGE;
+    iItem.pszText = data;
+    iItem.cchTextMax = lstrlenW(data);
+    iItem.iSubItem = iSubItem;
+    iItem.iItem = id;
+    iItem.iImage = id;
+
+    if (iSubItem)
+        return (ListView_SetItem(m_hWnd, &iItem) != -1);
+    else
+        return (ListView_InsertItem(m_hWnd, &iItem) != -1);
+}
+
+VOID TrayIconView::BuildIconList()
+{
+    HWND hToolbar = FindTrayToolbarWindowHandle();
+    int count = SendMessageW(hToolbar, TB_BUTTONCOUNT, 0, 0);
+
+    hImageList = ImageList_Create(16, 16, ILC_COLOR24, count, 0);
+
+    TrayIconData<TBBUTTON> IconData(LMEM_FIXED);
+    TBBUTTON tb = {0};
+    NOTIFYICONDATAW idata;
+
+    for(int idx =0 ; idx < count; idx++)
+    {
+        ::SendMessageW(hToolbar, TB_GETBUTTON, idx, (LPARAM)IconData.GetData());
+        IconData.ReadData(&tb);
+        IconData.ReadData<NOTIFYICONDATAW>(&idata,(PCHAR)tb.dwData);
+ 
+        HICON hIcon = CopyIcon(idata.hIcon);
+        if (hIcon)
+        {
+            ImageList_AddIcon(hImageList, hIcon);
+        }
+        InsertItem(idata.szTip, 0, idx);
+
+   }
+
+    SetImageList(hImageList, LVSIL_SMALL);
+}
+
+static TrayIconView *m_pListView = NULL;
+
 INT_PTR CALLBACK CustomizeNotifyIconsProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
 {
     switch(Message)
     {
         case WM_INITDIALOG:
+        {
+            m_pListView = new TrayIconView();
+            m_pListView->Create(hwnd, IDC_NOTIFICATION_LIST);
+            m_pListView->BuildIconList();
+            return TRUE;
+        }
 
-        return TRUE;
         case WM_COMMAND:
-           switch(LOWORD(wParam))
-           {
-               case IDOK:
-                   EndDialog(hwnd, IDOK);
-               break;
-               case IDCANCEL:
-                   EndDialog(hwnd, IDCANCEL);
-               break;
-           }
-       break;
-       default:
-           return FALSE;
+            switch(LOWORD(wParam))
+            {
+                case IDOK:
+                    EndDialog(hwnd, IDOK);
+                    break;
+
+                case IDCANCEL:
+                    EndDialog(hwnd, IDCANCEL);
+                break;
+            }
+            break;
+
+        case WM_CLOSE:
+            delete m_pListView;
```

Comment:
It's better to delete this in `WM_DESTROY`, but be sure to set it to `nullptr` after deleting it, so that a use-after-free will be noticable